lib/Conversion/AffineToHandshake/AffineToHandshake.cpp

```diff
@@ -0,0 +1,835 @@
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create the new handshake::FuncOp from the given mlir::FuncOp.
+    llvm::SmallVector<NamedAttribute, 4> attributes;
+    llvm::SmallVector<mlir::Type, 8> argTypes;
+    llvm::SmallVector<mlir::Type, 8> resTypes;
+
+    getFuncOpAttributes(funcOp, attributes);
+    getArgAndResultTypes(funcOp, argTypes, resTypes, rewriter);
+
+    auto funcType = rewriter.getFunctionType(argTypes, resTypes);
+    auto newFuncOp = rewriter.create<handshake::FuncOp>(
+        funcOp.getLoc(), funcOp.getName(), funcType, attributes);
+
+    // Move the body region from the mlir::FuncOp to the new
+    // handshake::FuncOp.
+    rewriter.inlineRegionBefore(funcOp.getBody(), newFuncOp.getBody(),
+                                newFuncOp.end());
+
+    // Explicitly convert the signature of the handshake::FuncOp body region.
+    unsigned numArguments = argTypes.size();
+    TypeConverter::SignatureConversion result(numArguments);
+    for (unsigned i = 0; i < numArguments; ++i)
+      result.addInputs(i, argTypes[i]);
+    rewriter.applySignatureConversion(&newFuncOp.getBody(), result);
+
+    // Now we start to rewrite the body of this new handshake::FuncOp.
+    // First of all, we create a pair of handshake::StartOp and
+    // handshake::ReturnOp in the newFuncOp.
+    setControlOnlyPath(newFuncOp, rewriter);
+    // Rewrite all the CallOp in the function body to handshake::InstanceOp.
+    rewriteCallOps(newFuncOp, rewriter);
+    // Rewrite memory operations
+    MemToAccess memToAccess;
+    rewriteMemoryOps(newFuncOp, memToAccess, rewriter);
+    // Create SinkOp for those dangling values.
+    addSinkOps(newFuncOp, rewriter);
+    // Rewrite ConstantOp with handshake::ConstantOp that has control connected.
+    rewriteConstantOps(newFuncOp, rewriter);
+    // There might be constant operations that are not being used.
+    addSinkOps(newFuncOp, rewriter);
+    // Make sure every result in the program has exactly one use.
+    addForkOps(newFuncOp, rewriter);
+
+    // Connect memory access operations to newly created handshake::MemoryOp.
+    connectAccessToMemory(newFuncOp, memToAccess, rewriter);
+
+    // Rewrite the StartOp and remove its duplicantions. A bit more on why there
+    // will be duplications: The single result of the StartOp will be used by
+    // many operations, e.g., handshake::ConstantOp, handshake::ReturnOp, etc.
+    // Changing their use to the block argument will avoid such problems.
+    rewriteStartOp(newFuncOp, rewriter);
+
+    // LLVM_DEBUG(newFuncOp.dump());
+
+    rewriter.eraseOp(funcOp);
+
+    return success();
+  }
+};
+
+struct HandshakePass
+    : public mlir::PassWrapper<HandshakePass, OperationPass<mlir::ModuleOp>> {
+  void runOnOperation() override {
+    ModuleOp m = getOperation();
+
+    ConversionTarget target(getContext());
+    target.addLegalDialect<HandshakeOpsDialect, StandardOpsDialect,
```

Comment:
I suppose this will eventually subsume and replace StandardToHandshake?

Reply:
I'm not sure whether this pass will replace StandardToHandshake. They do have many things in common, but there are quite a lot of differences, e.g., the conversion from `affine.for` that I will implement (some discussion above).